File: src/main/java/practice/user/application/in/UserFacade.java
package practice.user.application.in;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import practice.adaptor.in.web.dto.UserRequest;
import practice.account.application.in.CreateAccountUseCase;
import practice.user.domain.User;

@Service
@RequiredArgsConstructor
public class UserFacade {

  private final CreateAccountUseCase createAccountUseCase;
  private final RegisterUserUseCase registerUserUseCase;

  @Transactional
  public void createAccount(UserRequest userRequest) {
    String name = userRequest.getName();
    String email = userRequest.getEmail();

    if (registerUserUseCase.alreadyExists(name, email)) {
      return;
    }

    User user = registerUserUseCase.register(name, email);
    createAccountUseCase.createMainAccount(user.getId());
    createAccountUseCase.registerExternalAccount(user.getId(), userRequest.getExternalBank(), userRequest.getExternalAccountNumber());
  }

}

File: src/main/kotlin/practice/user/application/in/UserFacade.kt
package practice.user.application.`in`

import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import practice.account.application.`in`.CreateAccountUseCase
import practice.common.dto.UserRequest
import practice.user.domain.User

@Service
class UserFacade(
    private val createAccountUseCase: CreateAccountUseCase,
    private val registerUserUseCase: RegisterUserUseCase
) {

    @Transactional
    fun createAccount(userRequest: UserRequest) {
        val name: String = userRequest.name
        val email: String = userRequest.email

        if (registerUserUseCase.alreadyExists(name, email)) {
            return
        }

        val user: User = registerUserUseCase.register(name, email)
        createAccountUseCase.createMainAccount(user.id)
        createAccountUseCase.registerExternalAccount(
            user.id,
            userRequest.externalBank,
            userRequest.externalAccountNumber
        )
    }
}
